File: sdk/src/main/java/co/omise/android/ui/CreditCardActivity.java
package co.omise.android.ui;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;

import java.io.IOError;

import co.omise.android.CardIO;
import co.omise.android.CardNumber;
import co.omise.android.Client;
import co.omise.android.R;
import co.omise.android.TokenRequest;
import co.omise.android.TokenRequestListener;
import co.omise.android.models.APIError;
import co.omise.android.models.CardBrand;
import co.omise.android.models.Token;
import io.card.payment.CardIOActivity;
import io.card.payment.CreditCard;

public class CreditCardActivity extends Activity {
    // input
    public static final String EXTRA_PKEY = "CreditCardActivity.publicKey";
    public static final int REQUEST_CODE_CARD_IO = 1000;

    // output
    public static final int RESULT_OK = 100;
    public static final int RESULT_CANCEL = 200;

    public static final String EXTRA_TOKEN = "CreditCardActivity.token";
    public static final String EXTRA_TOKEN_OBJECT = "CreditCardActivity.tokenObject";
    public static final String EXTRA_CARD_OBJECT = "CreditCardActivity.cardObject";

    private Views views = new Views(this);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_credit_card);
        setTitle(R.string.default_form_title);

        views.spinner(R.id.spinner_expiry_month).setAdapter(new ExpiryMonthSpinnerAdapter());
        views.spinner(R.id.spinner_expiry_year).setAdapter(new ExpiryYearSpinnerAdapter());
        views.editText(R.id.edit_card_number).addTextChangedListener(new ActivityTextWatcher());
        views.button(R.id.button_submit).setOnClickListener(new ActivityOnClickListener());
    }

    @Override
    public void onBackPressed() {
        setResult(RESULT_CANCEL);
        super.onBackPressed();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_credit_card, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        menu.getItem(0).setVisible(CardIO.isAvailable() && views.button(R.id.button_submit).isEnabled());
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.menu_item_card_io) {
            if (CardIO.isAvailable()) {
                Intent intent = CardIO.buildIntent(this);
                startActivityForResult(intent, REQUEST_CODE_CARD_IO);
            }
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_CODE_CARD_IO) {
            if (data == null || !data.hasExtra(CardIOActivity.EXTRA_SCAN_RESULT)) {
                return;
            }

            CreditCard scanResult = data.getParcelableExtra(CardIOActivity.EXTRA_SCAN_RESULT);
            applyCardIOResult(scanResult);

        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    private class ActivityTextWatcher implements TextWatcher {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
        }

        @Override
        public void afterTextChanged(Editable s) {
            String pan = s.toString();
            if (pan.length() > 6) {
                CardBrand brand = CardNumber.brand(pan);
                if (brand != null && brand.getLogoResourceId() > -1) {
                    views.image(R.id.image_card_brand).setImageResource(brand.getLogoResourceId());
                    return;
                }
            }

            views.image(R.id.image_card_brand).setImageDrawable(null);
        }
    }

    private class ActivityOnClickListener implements View.OnClickListener {
        @Override
        public void onClick(View v) {
            if (v.getId() == R.id.button_submit) {
                submit();
            }
        }
    }

    private class ActivityTokenRequestListener implements TokenRequestListener {
        @Override
        public void onTokenRequestSucceed(TokenRequest request, Token token) {
            Intent data = new Intent();
            data.putExtra(EXTRA_TOKEN, token.id);
            data.putExtra(EXTRA_TOKEN_OBJECT, token);
            data.putExtra(EXTRA_CARD_OBJECT, token.card);

            setResult(RESULT_OK, data);
            finish();
        }

        @Override
        public void onTokenRequestFailed(TokenRequest request, Throwable throwable) {
            enableForm();

            TextView textView = views.textView(R.id.text_error_message);
            textView.setVisibility(View.VISIBLE);

            String message = null;
            if (throwable instanceof IOError) {
                message = getString(R.string.error_io, throwable.getMessage());
            } else if (throwable instanceof APIError) {
                message = getString(R.string.error_api, ((APIError) throwable).message);
            } else {
                message = getString(R.string.error_unknown, throwable.getMessage());
            }

            textView.setText(message);
        }
    }

    private void disableForm() {
        setFormEnabled(false);
    }

    private void enableForm() {
        setFormEnabled(true);
    }

    private void setFormEnabled(boolean enabled) {
        views.editText(R.id.edit_card_number).setEnabled(enabled);
        views.editText(R.id.edit_card_name).setEnabled(enabled);
        views.editText(R.id.edit_security_code).setEnabled(enabled);
        views.spinner(R.id.spinner_expiry_month).setEnabled(enabled);
        views.spinner(R.id.spinner_expiry_year).setEnabled(enabled);
        views.button(R.id.button_submit).setEnabled(enabled);
        invalidateOptionsMenu();
    }

    private void applyCardIOResult(CreditCard data) {
        EditText numberField = views.editText(R.id.edit_card_number);
        EditText nameField = views.editText(R.id.edit_card_name);
        EditText securityCodeField = views.editText(R.id.edit_security_code);

        if (data.cardNumber != null && !data.cardNumber.isEmpty()) {
            numberField.setText(CardNumber.format(data.cardNumber));
        }

        if (data.cardholderName != null && !data.cardholderName.isEmpty()) {
            nameField.setText(data.cardholderName);
        }

        if (data.isExpiryValid()) {
            Spinner spinner = views.spinner(R.id.spinner_expiry_month);
            ExpiryMonthSpinnerAdapter monthAdapter = (ExpiryMonthSpinnerAdapter) spinner.getAdapter();
            spinner.setSelection(monthAdapter.getPosition(data.expiryMonth));

            spinner = views.spinner(R.id.spinner_expiry_year);
            ExpiryYearSpinnerAdapter yearAdapter = (ExpiryYearSpinnerAdapter) spinner.getAdapter() ;
            spinner.setSelection(yearAdapter.getPosition(data.expiryYear));
        }

        if (data.cvv != null && !data.cvv.isEmpty()) {
            securityCodeField.setText(data.cvv);
        }

        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        if (numberField.getText() == null || numberField.getText().toString().isEmpty()) {
            numberField.requestFocus();
            imm.showSoftInput(numberField, InputMethodManager.SHOW_IMPLICIT);
        } else if (nameField.getText() == null || nameField.getText().toString().isEmpty()) {
            nameField.requestFocus();
            imm.showSoftInput(nameField, InputMethodManager.SHOW_IMPLICIT);
        } else if (securityCodeField.getText() == null || securityCodeField.getText().toString().isEmpty()) {
            securityCodeField.requestFocus();
            imm.showSoftInput(securityCodeField, InputMethodManager.SHOW_IMPLICIT);
        }
    }

    private void submit() {
        EditText numberField = views.editText(R.id.edit_card_number);
        EditText nameField = views.editText(R.id.edit_card_name);
        EditText securityCodeField = views.editText(R.id.edit_security_code);

        boolean valid = validateNonEmpty(numberField) &
                validateNonEmpty(nameField) &
                validateNonEmpty(securityCodeField) &
                validateLuhn(numberField);
        if (!valid) {
            return;
        }

        int expiryMonth = (int) views.spinner(R.id.spinner_expiry_month).getSelectedItem();
        int expiryYear = (int) views.spinner(R.id.spinner_expiry_year).getSelectedItem();

        TokenRequest tokenRequest = new TokenRequest();
        tokenRequest.number = numberField.getText().toString();
        tokenRequest.name = nameField.getText().toString();
        tokenRequest.securityCode = securityCodeField.getText().toString();
        tokenRequest.expirationMonth = expiryMonth;
        tokenRequest.expirationYear = expiryYear;

        disableForm();

        String pkey = getIntent().getStringExtra(EXTRA_PKEY);
        ActivityTokenRequestListener listener = new ActivityTokenRequestListener();
        try {
            new Client(pkey).send(tokenRequest, listener);
        } catch (Exception ex) {
            listener.onTokenRequestFailed(tokenRequest, ex);
        }

    }

    private boolean validateNonEmpty(EditText field) {
        String value = field.getText().toString().trim();
        if (value.isEmpty()) {
            field.setError(String.format(getString(R.string.error_required), field.getHint()));
            return false;
        }

        return true;
    }

    private boolean validateLuhn(EditText field) {
        String value = field.getText().toString().trim();
        if (!CardNumber.luhn(value)) {
            field.setError(String.format(getString(R.string.error_invalid), field.getHint()));
            return false;
        }

        return true;
    }
}

File: sdk/src/main/java/co/omise/android/ui/CreditCardActivity.kt
package co.omise.android.ui

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.WindowManager
import android.widget.EditText
import android.widget.TextView
import androidx.activity.OnBackPressedCallback
import co.omise.android.CardNumber
import co.omise.android.R
import co.omise.android.api.Client
import co.omise.android.api.Request
import co.omise.android.api.RequestListener
import co.omise.android.databinding.ActivityCreditCardBinding
import co.omise.android.extensions.getMessageFromResources
import co.omise.android.extensions.parcelable
import co.omise.android.extensions.setOnAfterTextChangeListener
import co.omise.android.extensions.setOnClickListener
import co.omise.android.extensions.textOrNull
import co.omise.android.models.APIError
import co.omise.android.models.BackendType
import co.omise.android.models.Capability
import co.omise.android.models.CardHolderDataField
import co.omise.android.models.CardHolderDataList
import co.omise.android.models.CardParam
import co.omise.android.models.CountryInfo
import co.omise.android.models.PaymentMethod
import co.omise.android.models.Source
import co.omise.android.models.Token
import co.omise.android.models.backendType
import com.google.android.material.snackbar.Snackbar
import org.jetbrains.annotations.TestOnly
import java.io.IOError
import java.util.Locale

/**
 * CreditCardActivity is the UI class for taking credit card information input from the user.
 */
class CreditCardActivity : OmiseActivity() {
    private lateinit var pKey: String
    private lateinit var cardHolderData: CardHolderDataList
    private lateinit var client: Client
    private lateinit var binding: ActivityCreditCardBinding

    /**
     * Target countries that supports AVS or the Address Verification System.
     * @see [link](https://www.omise.co/How-to-improve-my-authorization-rate-for-US-UK-and-Canadian-cardholders)
     */
    private val avsCountries = CountryInfo.ALL.filter { listOf("US", "GB", "CA").contains(it.code) }

    private val editTexts: Map<OmiseEditText, TextView> by lazy {
        mapOf(
            binding.editCardNumber to binding.textCardNumberError,
            binding.editCardName to binding.textCardNameError,
            binding.editExpiryDate to binding.textExpiryDateError,
            binding.editSecurityCode to binding.textSecurityCodeError,
            binding.editCountry to binding.textCountryError,
            binding.editStreet1 to binding.textStreet1Error,
            binding.editCity to binding.textCityError,
            binding.editState to binding.textStateError,
            binding.editPostalCode to binding.textPostalCodeError,
            binding.editEmail to binding.textEmailError,
            binding.editPhoneNumber to binding.textPhoneNumberError,
        )
    }

    private val billingAddressEditTexts: Map<OmiseEditText, TextView> by lazy {
        mapOf(
            binding.editCountry to binding.textCountryError,
            binding.editStreet1 to binding.textStreet1Error,
            binding.editCity to binding.textCityError,
            binding.editState to binding.textStateError,
            binding.editPostalCode to binding.textPostalCodeError,
        )
    }

    private var selectedCountry: CountryInfo? = null
        set(value) {
            field = value
            value?.let {
                invalidateBillingAddressForm()
            }
        }

    private fun isEmailValid(emailEdit: OmiseEditText): Boolean {
        return android.util.Patterns.EMAIL_ADDRESS.matcher(emailEdit.text!!).matches()
    }

    private fun isPhoneNumberValid(phoneNumberEdit: OmiseEditText): Boolean {
        return android.util.Patterns.PHONE.matcher(phoneNumberEdit.text!!).matches()
    }

    private fun updateEmailErrorText(hasFocus: Boolean) {
        // Clear error when field has focus (consistent with other fields)
        if (hasFocus) {
            with(binding.textEmailError) {
                text = ""
                visibility = GONE
            }
            return
        }

        // When field loses focus, validate only if not empty
        if (binding.editEmail.text?.isEmpty() == true || isEmailValid(binding.editEmail)) {
            with(binding.textEmailError) {
                text = ""
                visibility = GONE
            }
            return
        }

        binding.textEmailError.visibility = VISIBLE
        binding.textEmailError.text = getString(R.string.error_invalid_email)
    }

    private fun updatePhoneErrorText(hasFocus: Boolean) {
        // Clear error when field has focus (consistent with other fields)
        if (hasFocus) {
            with(binding.textPhoneNumberError) {
                text = ""
                visibility = GONE
            }
            return
        }

        // When field loses focus, validate only if not empty
        if (binding.editPhoneNumber.text?.isEmpty() == true || isPhoneNumberValid(binding.editPhoneNumber)) {
            with(binding.textPhoneNumberError) {
                text = ""
                visibility = GONE
            }
            return
        }

        binding.textPhoneNumberError.visibility = VISIBLE
        binding.textPhoneNumberError.text = getString(R.string.error_invalid_phone_number)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (intent.getBooleanExtra(EXTRA_IS_SECURE, true)) {
            window.addFlags(WindowManager.LayoutParams.FLAG_SECURE)
        }

        binding = ActivityCreditCardBinding.inflate(layoutInflater)
        setContentView(binding.root)

        require(intent.hasExtra(EXTRA_PKEY)) { "Could not find ${::EXTRA_PKEY.name}." }
        pKey = requireNotNull(intent.getStringExtra(EXTRA_PKEY)) { "${::EXTRA_PKEY.name} must not be null." }
        cardHolderData = intent.parcelable<CardHolderDataList>(EXTRA_CARD_HOLDER_DATA) ?: CardHolderDataList(arrayListOf())

        if (!this::client.isInitialized) {
            client = Client(pKey)
        }
        val onBackPressedCallback =
            object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    setResult(RESULT_CANCELED)
                    finish()
                }
            }

        onBackPressedDispatcher.addCallback(this, onBackPressedCallback)

        initialize()
    }

    private fun EditText.getErrorMessage(): String? {
        return when (this) {
            binding.editCardNumber -> getString(R.string.error_invalid_card_number)
            binding.editCardName -> getString(R.string.error_invalid_card_name)
            binding.editExpiryDate -> getString(R.string.error_invalid_expiration_date)
            binding.editSecurityCode -> getString(R.string.error_invalid_security_code)
            binding.editStreet1 -> getString(R.string.error_required_street1)
            binding.editCity -> getString(R.string.error_required_city)
            binding.editState -> getString(R.string.error_required_state)
            binding.editPostalCode -> getString(R.string.error_required_postal_code)
            binding.editEmail -> getString(R.string.error_invalid_email)
            binding.editPhoneNumber -> getString(R.string.error_invalid_phone_number)
            else -> null
        }
    }

    private fun initialize() {
        setTitle(R.string.default_form_title)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        binding.buttonSubmit.setOnClickListener(::submit)
        binding.buttonSecurityCodeTooltip.setOnClickListener(::showSecurityCodeTooltipDialog)
        binding.editCountry.setOnClickListener(::showCountryDropdownDialog)
        cardHolderDataVisibility()

        editTexts.forEach { (editText, errorText) ->
            editText.setOnFocusChangeListener { _, hasFocus ->
                // Handle card holder data fields (email and phone) with their special logic
                if (editText == binding.editEmail && cardHolderData.fields.contains(CardHolderDataField.EMAIL)) {
                    updateEmailErrorText(hasFocus)
                } else if (editText == binding.editPhoneNumber && cardHolderData.fields.contains(CardHolderDataField.PHONE_NUMBER)) {
                    updatePhoneErrorText(hasFocus)
                } else if (editText != binding.editEmail && editText != binding.editPhoneNumber) {
                    // Handle regular fields - skip email and phone as they have special handling
                    if (!hasFocus) {
                        try {
                            editText.validate()
                            errorText.text = null
                            errorText.visibility = GONE
                        } catch (e: InputValidationException.InvalidInputException) {
                            errorText.text = editText.getErrorMessage()
                            errorText.visibility = VISIBLE
                        } catch (e: InputValidationException.EmptyInputException) {
                            if (isBillingAddressRequired() && billingAddressEditTexts.containsKey(editText)) {
                                errorText.text = editText.getErrorMessage()
                                errorText.visibility = VISIBLE
                            } else {
                                errorText.text = null
                                errorText.visibility = GONE
                            }
                        }
                    } else {
                        // Clear error text when gaining focus
                        errorText.text = null
                        errorText.visibility = GONE
                    }
                }
                updateSubmitButton()
            }
            editText.setOnAfterTextChangeListener(::updateSubmitButton)
        }

        invalidateBillingAddressForm()

        getCapability()
    }

    @TestOnly
    fun setClient(client: Client) {
        this.client = client
    }

    private fun getCapability() {
        val getCapabilityRequest = Capability.GetCapabilitiesRequestBuilder().build()
        client.send(
            getCapabilityRequest,
            object : RequestListener<Capability> {
                override fun onRequestSucceed(model: Capability) {
                    val countryCode = model.country ?: Locale.getDefault().country
                    selectedCountry = CountryInfo.ALL.find { it.code == countryCode }
                }

                override fun onRequestFailed(throwable: Throwable) {
                    Snackbar.make(binding.scrollview, throwable.message.toString(), Snackbar.LENGTH_LONG).show()
                }
            },
        )
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                setResult(RESULT_CANCELED)
                finish()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun disableForm() {
        setFormEnabled(false)
    }

    private fun enableForm() {
        setFormEnabled(true)
    }

    private fun setFormEnabled(enabled: Boolean) {
        editTexts.forEach { (editText, _) -> editText.isEnabled = enabled }
        binding.buttonSubmit.isEnabled = enabled
    }

    private fun handleRequestFailed(throwable: Throwable) {
        enableForm()

        val message =
            when (throwable) {
                is IOError -> getString(R.string.error_io, throwable.message)
                is APIError -> throwable.getMessageFromResources(resources)
                else -> getString(R.string.error_unknown, throwable.message)
            }

        Snackbar.make(binding.scrollview, message, Snackbar.LENGTH_LONG).show()
    }

    private fun submit() {
        disableForm()

        val cardParam =
            CardParam(
                name = binding.editCardName.cardName,
                number = binding.editCardNumber.cardNumber,
                expirationMonth = binding.editExpiryDate.expiryMonth,
                expirationYear = binding.editExpiryDate.expiryYear,
                securityCode = binding.editSecurityCode.securityCode,
                country = selectedCountry?.code,
                street1 = binding.editStreet1.textOrNull?.toString(),
                city = binding.editCity.textOrNull?.toString(),
                state = binding.editState.textOrNull?.toString(),
                postalCode = binding.editPostalCode.textOrNull?.toString(),
                email = binding.editEmail.textOrNull?.toString(),
                phoneNumber = binding.editPhoneNumber.textOrNull?.toString(),
            )

        val request = Token.CreateTokenRequestBuilder(cardParam).build()
        val installmentTerms = intent.getIntExtra(EXTRA_SELECTED_INSTALLMENTS_TERM, 0)
        // check if the user is coming from a WLB installment
        var sourceRequest: Request<Source>? = null
        if (installmentTerms != 0) {
            // create a source request
            val amount = requireNotNull(intent.getLongExtra(EXTRA_AMOUNT, 0)) { "${::EXTRA_AMOUNT.name} must not be null." }

            val currency = requireNotNull(intent.getStringExtra(EXTRA_CURRENCY)) { "${::EXTRA_CURRENCY.name} must not be null." }
            val paymentMethod =
                requireNotNull(
                    intent.parcelable<PaymentMethod>(EXTRA_SELECTED_INSTALLMENTS_PAYMENT_METHOD),
                ) {
                    "${::EXTRA_SELECTED_INSTALLMENTS_PAYMENT_METHOD.name} must not be null."
                }
            val sourceType = (paymentMethod.backendType as? BackendType.Source)?.sourceType ?: return
            val capability =
                requireNotNull(
                    intent.parcelable<Capability>(EXTRA_CAPABILITY),
                ) { "${::EXTRA_CAPABILITY.name} must not be null." }
            sourceRequest =
                Source.CreateSourceRequestBuilder(amount, currency, sourceType)
                    .installmentTerm(installmentTerms)
                    .zeroInterestInstallments(capability.zeroInterestInstallments)
                    .build()
        }
        client.send(
            request,
            object : RequestListener<Token> {
                override fun onRequestSucceed(model: Token) {
                    val data =
                        Intent().apply {
                            putExtra(EXTRA_TOKEN, model.id)
                            putExtra(EXTRA_TOKEN_OBJECT, model)
                            putExtra(EXTRA_CARD_OBJECT, model.card)
                        }
                    if (sourceRequest == null) {
                        setResult(Activity.RESULT_OK, data)
                        finish()
                    } else {
                        // create source request
                        client.send(
                            sourceRequest,
                            object : RequestListener<Source> {
                                override fun onRequestSucceed(model: Source) {
                                    data.apply {
                                        putExtra(EXTRA_SOURCE_OBJECT, model)
                                    }
                                    setResult(Activity.RESULT_OK, data)
                                    finish()
                                }

                                override fun onRequestFailed(throwable: Throwable) {
                                    handleRequestFailed(throwable)
                                }
                            },
                        )
                    }
                }

                override fun onRequestFailed(throwable: Throwable) {
                    handleRequestFailed(throwable)
                }
            },
        )
    }

    private fun updateSubmitButton() {
        val validationResults = mutableMapOf<OmiseEditText, Boolean>()

        editTexts.keys.forEach { editText ->
            when {
                // Required fields (card info)
                editText in listOf(binding.editCardNumber, binding.editCardName, binding.editExpiryDate, binding.editSecurityCode) -> {
                    validationResults[editText] = editText.isValid
                }
                // Billing address fields (only required for AVS countries)
                editText in billingAddressEditTexts -> {
                    if (isBillingAddressRequired()) {
                        validationResults[editText] = editText.isValid
                    }
                    // If billing address not required, these fields don't affect validation
                }
                // Email field (required when requested by merchant)
                editText == binding.editEmail && cardHolderData.fields.contains(CardHolderDataField.EMAIL) -> {
                    validationResults[editText] = isEmailValid(binding.editEmail)
                }
                // Phone number field (required when requested by merchant)
                editText == binding.editPhoneNumber && cardHolderData.fields.contains(CardHolderDataField.PHONE_NUMBER) -> {
                    validationResults[editText] = isPhoneNumberValid(binding.editPhoneNumber)
                }
                // Other fields that are not visible/required don't affect validation
            }
        }

        // All included fields must be valid
        val isFormValid = validationResults.values.all { it }
        binding.buttonSubmit.isEnabled = isFormValid
    }

    private fun showSecurityCodeTooltipDialog() {
        val brand = CardNumber.brand(binding.editCardNumber.cardNumber)
        val dialog = SecurityCodeTooltipDialogFragment.newInstant(brand)
        dialog.show(supportFragmentManager, null)
    }

    private fun showCountryDropdownDialog() {
        val dialog = CountryListDialogFragment()
        dialog.listener =
            object : CountryListDialogFragment.CountryListDialogListener {
                override fun onCountrySelected(country: CountryInfo) {
                    selectedCountry = country
                }
            }
        dialog.show(supportFragmentManager, null)
    }

    private fun invalidateBillingAddressForm() {
        binding.editCountry.setText(selectedCountry?.name)
        binding.billingAddressContainer.visibility = if (isBillingAddressRequired()) View.VISIBLE else View.GONE
        billingAddressEditTexts.forEach { (editText, errorText) ->
            if (editText != binding.editCountry) {
                editText.text = null
                errorText.text = null
            }
        }
        updateSubmitButton()
    }

    private fun cardHolderDataVisibility() {
        if (cardHolderData.fields.contains(CardHolderDataField.EMAIL)) {
            binding.editEmail.visibility = View.VISIBLE
            binding.textEmailTitle.visibility = View.VISIBLE
        } else {
            binding.editEmail.visibility = View.GONE
            binding.textEmailTitle.visibility = View.GONE
            binding.textEmailError.visibility = View.GONE
        }

        if (cardHolderData.fields.contains(CardHolderDataField.PHONE_NUMBER)) {
            binding.editPhoneNumber.visibility = View.VISIBLE
            binding.textPhoneNumberTitle.visibility = View.VISIBLE
        } else {
            binding.editPhoneNumber.visibility = View.GONE
            binding.textPhoneNumberTitle.visibility = View.GONE
            binding.textPhoneNumberError.visibility = View.GONE
        }
    }

    private fun isBillingAddressRequired(): Boolean {
        return selectedCountry != null && avsCountries.contains(selectedCountry)
    }
}
